### PR TITLE
GH984 Add overload for DataFrame.clip and update those for Series.clip

### LINKED ISSUE #984: Create overloads for `DataFrame.clip()`
From discussion in #982 .  See all comments there.

              It's possible (but not required) that you could also handle this via typing checks by creating overloads that are the acceptable combinations of the arguments `lower`, `upper` and `axis`

_Originally posted by @Dr-Irv in https://github.com/pandas-dev/pandas-stubs/pull/982#discussion_r1725010182_

### PR DESCRIPTION
**Pull request by loicdiridollou**
- [x] Closes #984 
- [x] Tests added: Please use `assert_type()` to assert the type of any return value

Turns out there is a nasty edge case when you pass everything as None for upper and lower and ask for `inplace=True`, it will return `Self` and not `None`.

### CODE REVIEW

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
can you add tests where either `lower` or `upper` is not specified?

I think that might lead to some changes in how you wrote the overloads.

**Comment by loicdiridollou** [comment]
Turns out no changes were needed but agreed that this time we check all the cases.

**Comment by Dr-Irv** [comment]
I'm a little confused here, as you closed the PR and then reopened it.  Should I review?

**Comment by loicdiridollou** [comment]
Yes it is open for review, I have accidentally pressed the close button, my
mistake.

File: tests/test_frame.py [hunk 1]
```diff
@@ -778,6 +778,117 @@ def test_dataframe_clip() -> None:
...
+        assert_type(df.clip(upper=15, axis="index", inplace=True), None),
+        type(None),
+    )
+    check(
+        assert_type(df.clip(upper=pd.Series([1, 2]), axis="index", inplace=True), None),
+        type(None),
+    )
+    check(assert_type(df.clip(upper=None, axis="index"), pd.DataFrame), pd.DataFrame)
+    check(assert_type(df.clip(upper=15, axis="index"), pd.DataFrame), pd.DataFrame)
+    check(assert_type(df.clip(upper=None, axis="index"), pd.DataFrame), pd.DataFrame)
```
Comment (Dr-Irv) [inline tests/test_frame.py:810]: lines 808 and 810 are the same

File: tests/test_frame.py [hunk 2]
```diff
@@ -778,6 +778,117 @@ def test_dataframe_clip() -> None:
...
+        pd.DataFrame,
+    )
+    check(
+        assert_type(df.clip(lower=pd.Series([1, 2]), axis="index"), pd.DataFrame),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(df.clip(lower=None, axis="index"), pd.DataFrame),
+        pd.DataFrame,
+    )
```
Comment (Dr-Irv) [inline tests/test_frame.py:881]: lines 863 and 871 and 879  are the same test

File: tests/test_frame.py [hunk 3]
```diff
@@ -659,23 +659,235 @@ def test_types_quantile() -> None:
...
+    check(
+        assert_type(df.clip(lower=5, axis="index", inplace=True), None),
+        type(None),
+    )
+    check(
+        assert_type(df.clip(lower=pd.Series([1, 2]), axis="index", inplace=True), None),
+        type(None),
+    )
+    check(
+        assert_type(df.clip(lower=None, axis="index", inplace=True), pd.DataFrame),
```
Comment (Dr-Irv) [inline tests/test_frame.py:859]: lines 847 and 859 are the same test

File: tests/test_frame.py [hunk 4]
```diff
@@ -659,23 +659,235 @@ def test_types_quantile() -> None:
...
+        assert_type(df.clip(lower=None, axis="index"), pd.DataFrame),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(df.clip(lower=None, axis=0, inplace=True), pd.DataFrame),
+        pd.DataFrame,
+    )
+    check(assert_type(df.clip(lower=5, axis=0, inplace=True), None), type(None))
+    check(
+        assert_type(df.clip(lower=None, axis=0, inplace=True), pd.DataFrame),
```
Comment (Dr-Irv) [inline tests/test_frame.py:888]: lines 883 and 888 are the same test

File: tests/test_frame.py [hunk 5]
```diff
@@ -659,23 +659,235 @@ def test_types_quantile() -> None:
...
+        assert_type(df.clip(lower=None, axis=None, inplace=True), pd.DataFrame),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(df.clip(lower=5, axis=None, inplace=True), None),
+        type(None),
+    )
+    check(
+        assert_type(df.clip(lower=None, axis=None, inplace=True), pd.DataFrame),
+        pd.DataFrame,
```
Comment (Dr-Irv) [inline tests/test_frame.py:837]: lines 828 and 836 are the same test

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
I think you may also need similar overloads for `clip` in `series.pyi` and then similar tests that omit the `lower` or `upper` arguments.

**Comment by loicdiridollou** [comment]
Since clip for series does not use the axis when passing a list like
argument to lower or upper the overloads are already there. Let me clean up
the tests and add for those without upper or lower passed and we should be
in a good place, the big work was on the data frame side.

**Comment by loicdiridollou** [comment]
Turns out the axis parameter does not need anything specific (to split the overloads) since it is not used (unless you pass something not allowed).

**Review by Dr-Irv — APPROVED** [review]
thanks @loicdiridollou
